core/tx_list.go

```diff
@@ -464,121 +468,96 @@ func (l *txPricedList) Put(tx *types.Transaction) {
 func (l *txPricedList) Removed(count int) {
 	// Bump the stale counter, but exit if still too low (< 25%)
 	l.stales += count
-	if l.stales <= len(*l.items)/4 {
+	if l.stales <= len(*l.remotes)/4 {
 		return
 	}
 	// Seems we've reached a critical number of stale transactions, reheap
-	reheap := make(priceHeap, 0, l.all.Count())
-
-	l.stales, l.items = 0, &reheap
-	l.all.Range(func(hash common.Hash, tx *types.Transaction) bool {
-		*l.items = append(*l.items, tx)
-		return true
-	})
-	heap.Init(l.items)
+	l.Rebuild()
 }
 
 // Cap finds all the transactions below the given price threshold, drops them
 // from the priced list and returns them for further removal from the entire pool.
-func (l *txPricedList) Cap(threshold *big.Int, local *accountSet) types.Transactions {
+//
+// Note: only remote transactions will be considered for eviction.
+func (l *txPricedList) Cap(threshold *big.Int) types.Transactions {
 	drop := make(types.Transactions, 0, 128) // Remote underpriced transactions to drop
-	save := make(types.Transactions, 0, 64)  // Local underpriced transactions to keep
-
-	for len(*l.items) > 0 {
+	for len(*l.remotes) > 0 {
 		// Discard stale transactions if found during cleanup
-		tx := heap.Pop(l.items).(*types.Transaction)
-		if l.all.Get(tx.Hash()) == nil {
+		tx := heap.Pop(l.remotes).(*types.Transaction)
+		if l.all.GetRemote(tx.Hash()) == nil { // Removed or migrated
 			l.stales--
 			continue
 		}
 		// Stop the discards if we've reached the threshold
 		if tx.GasPriceIntCmp(threshold) >= 0 {
-			save = append(save, tx)
+			heap.Push(l.remotes, tx)
```

Comment:
I'm not sure it makes any difference, but might be worth benchmarking: doing a `heap.Peek` instead of `Pop + Push`. ?

Reply:
Yes I will do a benchmark for it

Reply:
Well, `heap` has no `Peak` function...

Reply:
> guerilla peek

That's not a trick actually, that's a guaranteed API, alas looks a bit odd:

> The minimum element in the tree is the root, at index 0.
> https://golang.org/pkg/container/heap/

Re Pop/Push vs Peek, each Pop and Push on a heap is log(n). So if we can peek, it's always much better.

Reply:
done